### DISCUSSION TITLE
BelongsToManyMultiSelect - choose from dropdown

### DISCUSSION THREAD

**Discussion by vv-ac** [filamentphp/filament#2347] [Help] [answered]
Hi all,

I use numerous multi selects and they work fine.
THe only thing bugging me is that when I have a BelongsToMaynMultiSelect and I click in it, I get the following:
![grafik](https://user-images.githubusercontent.com/67121752/167637132-79d1c8cb-d26d-4a5f-a05e-2b3673a17517.png)

Now, I can start typing and chose single elements from the search results.
Do I miss sth which makes this the only option?
I would like to see all possible results (or limited to ~50), so a user does not need to memorize the name of elements to choose.

I see, that it is planned to have choices.js to be implemented which seems to have that:
![grafik](https://user-images.githubusercontent.com/67121752/167638005-28736258-9dd1-4ec2-bdc0-dd9a851ff036.png)

I appreciate every remark.
Thanks!

**Comment by danharrin** [comment, marked as answer]
MultiSelect does this by default. For BelongsToManyMultiSelect, you need to `->preload()` - we don't do this by default since often there are many options and it can result in performance issues.

Choices.js was released in v2.12.0

**Reply by vv-ac** [reply]
Thanks!
This one though does not preload the options:
```
MultiSelect::make('requirement_source_types')
                    ->getSearchResultsUsing(fn (string $searchQuery) => Resource::where('name', 'like', "%{$searchQuery}%")->limit(50)->pluck('name', 'id'))
                    ->getOptionLabelsUsing(fn (array $values) => Resourcefind($values)->pluck('name'))
                    ->reactive()
                    ->columnSpan(['default' => 1])
```

->preload() is unknown to MultiSelect afaik.

Isn't it working when using a custom query?

**Reply by danharrin** [reply]
If you're using MultiSelect instead of BelongsToManyMultiSelect, you need to manually pass `->options()`. Check out the docs

**Reply by vv-ac** [reply]
I did and it said
```
If you have lots of options and want to populate them based on a database search or other external data source, you can use the getSearchResultsUsing() and getOptionLabelsUsing() methods instead of options().
```

I will re-read it.

**Reply by danharrin** [reply]
Yeah, so you can use `->getSearchResultsUsing()` in addition to `->options()` if you want to have AJAX search as well as some preloaded options. BelongsToManyMultiSelect handles this with the `->preload()` method, which means you dont need to manually define the options or search results.

**Reply by vv-ac** [reply]
Thank you for helping once more. I am still new to it so please bear with me asking simple questions here and there :-).

Everyone interested, add that:
```
->options(Resource::all()->pluck('name', 'id')->toArray())
```
